src/Compilers/CSharp/Test/Emit2/Semantics/CollectionLiteralTests.cs

```diff
@@ -5644,6 +5667,2548 @@ private static void VerifyTypes(SemanticModel model, ExpressionSyntax expr, stri
+                using System.Runtime.CompilerServices;
+                public class Container<T>
+                {
+                    [CollectionBuilder(typeof(MyCollectionBuilder), "Create")]
+                    public struct MyCollection<U> : IEnumerable<U>
+                    {
+                        private readonly List<U> _list;
+                        public MyCollection(List<U> list) { _list = list; }
+                        public IEnumerator<U> GetEnumerator() => _list.GetEnumerator();
+                        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+                    }
+                }
+                public sealed class MyCollectionBuilder
+                {
+                    public static Container<T>.MyCollection<U> Create<T, U>(ReadOnlySpan<U> items)
+                        => new Container<T>.MyCollection<U>(new List<U>(items.ToArray()));
+                }
+                """;
+            var comp = CreateCompilation(new[] { sourceA, s_constructibleCollectionAttributeDefinition }, targetFramework: TargetFramework.Net70);
+            var refA = AsReference(comp, useCompilationReference);
+
+            string sourceB = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        Container<int>.MyCollection<string> x = [];
+                        x.Report(includeType: true);
+                        Container<string>.MyCollection<int> y = [1, 2, 3];
+                        y.Report(includeType: true);
+                    }
+                }
+                """;
+            CompileAndVerify(new[] { sourceB, s_collectionExtensions }, references: new[] { refA }, targetFramework: TargetFramework.Net70, expectedOutput: "(Container<T>.MyCollection<System.Int32, System.String>) [], (Container<T>.MyCollection<System.String, System.Int32>) [1, 2, 3], ");
+        }
+
+        [CombinatorialData]
+        [ConditionalTheory(typeof(CoreClrOnly))]
+        public void CollectionBuilder_GenericType_ElementTypeFirstOfTwo(bool useCompilationReference)
+        {
+            string sourceA = """
+                using System;
+                using System.Collections;
+                using System.Collections.Generic;
+                using System.Runtime.CompilerServices;
+                [CollectionBuilder(typeof(MyCollectionBuilder), "Create")]
+                public struct MyCollection<T, U> : IEnumerable<T>
+                {
+                    private readonly List<T> _list;
+                    public MyCollection(List<T> list) { _list = list; }
+                    public IEnumerator<T> GetEnumerator() => _list.GetEnumerator();
+                    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+                }
+                public sealed class MyCollectionBuilder
+                {
+                    public static MyCollection<T, U> Create<T, U>(ReadOnlySpan<T> items)
```

Comment:
fascinating.  i didn't realize this woudl work.  So we do this not through inference from the arg, but by direct passing of the resultant collection type args to the method.   

IIRC, this is what we decided in LDM right?  I just didn't realize the implication until now :)

Reply:
Yes, we use the collection type arguments for the builder. I don't know if we covered this explicitly at LDM, but I will update the proposal to cover this.